pennylane/hf/tapering.py

```diff
@@ -0,0 +1,207 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This module contains the functions needed for tapering qubits.
+"""
+# pylint: disable=unnecessary-lambda
+import functools
+import itertools
+
+import autograd.numpy as anp
+import pennylane as qml
+from pennylane import numpy as np
+
+
+def _observable_mult(obs_a, obs_b):
+    r"""Multiply two PennyLane observables together.
+
+    Each observable should be a linear combination of Pauli words, e.g.,
+    :math:`\sum_{k=0}^{N} c_k O_k`, and represented as a PennyLane Hamiltonian.
+
+    Args:
+        obs_a (Hamiltonian): first observable
+        obs_b (Hamiltonian): second observable
+
+    Returns:
+        .Hamiltonian: Observable expressed as a PennyLane Hamiltonian
+
+    **Example**
+
+    >>> c = np.array([0.5, 0.5])
+    >>> obs_a = qml.Hamiltonian(c, [qml.PauliX(0) @ qml.PauliY(1), qml.PauliX(0) @ qml.PauliZ(1)])
+    >>> obs_b = qml.Hamiltonian(c, [qml.PauliX(0) @ qml.PauliX(1), qml.PauliZ(0) @ qml.PauliZ(1)])
+    >>> print(_observable_mult(obs_a, obs_b))
+      (-0.25j) [Z1]
+    + (-0.25j) [Y0]
+    + ( 0.25j) [Y1]
+    + ((0.25+0j)) [Y0 X1]
+    """
+    o = []
+    c = []
+    for i in range(len(obs_a.terms[0])):
+        for j in range(len(obs_b.terms[0])):
+            op, phase = qml.grouping.pauli_mult_with_phase(obs_a.terms[1][i], obs_b.terms[1][j])
+            o.append(op)
+            c.append(phase * obs_a.terms[0][i] * obs_b.terms[0][j])
+
+    return _simplify(qml.Hamiltonian(qml.math.stack(c), o))
+
+
+def _simplify(h, cutoff=1.0e-12):
+    r"""Add together identical terms in the Hamiltonian.
+
+    The Hamiltonian terms with identical Pauli words are added together and eliminated if the
+    overall coefficient is zero.
+
+    Args:
+        h (Hamiltonian): PennyLane Hamiltonian
+
+    Returns:
+        .Hamiltonian: Simplified PennyLane Hamiltonian
+
+    **Example**
+
+    >>> c = np.array([0.5, 0.5])
+    >>> h = qml.Hamiltonian(c, [qml.PauliX(0) @ qml.PauliY(1), qml.PauliX(0) @ qml.PauliY(1)])
+    >>> print(_simplify(h))
+    (1.0) [X0 Y1]
+    """
+    s = []
+    w = list(range(max(h.wires.tolist()) + 1))
+    wiremap = dict(zip(w, w))
+    for term in h.terms[1]:
+        term = qml.operation.Tensor(term).prune()
+        s.append(qml.grouping.pauli_word_to_string(term, wire_map=wiremap))
+
+    o = list(set(s))
+    c = [0.0] * len(o)
+    for i, item in enumerate(s):
+        c[o.index(item)] += h.terms[0][i]
+    c = qml.math.stack(c)
+
+    coeffs = []
+    ops = []
+    nonzero_ind = np.argwhere(abs(c) > cutoff).flatten()
+    for i in nonzero_ind:
+        coeffs.append(c[i])
+        ops.append(qml.grouping.string_to_pauli_word(o[i]))
+
+    if coeffs:
+        coeffs = qml.math.stack(coeffs)
```

Comment:
This is a fine way of doing it, but I think it hides some of the logic we have in place to handle empty lists. This suggestion is just to promote a more transparent way of coding 👍🏼 
```suggestion
    try:
        coeffs = qml.math.stack(coeffs)
    except ValueError:
        pass
```